src/main/java/com/uid2/admin/vertx/service/ClientSideKeypairService.java

```diff
@@ -0,0 +1,250 @@
+                return;
+            }
+        } else {
+            normalizedContactEmail = keypair.getContact();
+        }
+        if(disabled == null) {
+            disabled = keypair.isDisabled();
+        }
+
+        final ClientSideKeypair newKeypair = new ClientSideKeypair(
+                keypair.getSubscriptionId(),
+                keypair.encodePublicKeyToString(),
+                keypair.encodePrivateKeyToString(),
+                keypair.getSiteId(),
+                normalizedContactEmail,
+                keypair.getCreated(),
+                disabled);
+
+
+        Set<ClientSideKeypair> allKeypairs = new HashSet<>(this.keypairStore.getAll());
+        allKeypairs.remove(keypair);
+        allKeypairs.add(newKeypair);
+        try {
+            storeWriter.upload(allKeypairs, null);
+        }  catch (Exception e) {
+            rc.fail(500, e);
+            return;
+        }
+
+        final JsonObject json = toJson(newKeypair);
+        rc.response().putHeader(HttpHeaders.CONTENT_TYPE, "application/json")
+                .end(json.encode());
+    }
+
+    private void handleListAllKeypairs(RoutingContext rc) {
+        try {
+            final JsonArray ja = new JsonArray();
+            this.keypairStore.getSnapshot().getAll().forEach(k -> ja.add(toJson(k)));
+            rc.response()
+                    .putHeader(HttpHeaders.CONTENT_TYPE, "application/json")
+                    .end(ja.encode());
+        } catch (Exception e) {
+            rc.fail(500, e);
+        }
+    }
+
+    private void handleListKeypair(RoutingContext rc) {
+        String subscriptionId;
+        try {
+            subscriptionId = rc.pathParam("subscriptionId");
+        } catch (Exception e) {
+            ResponseUtil.error(rc, 400, "Failed to parse subscription id from request path");
+            return;
+        }
+
+        ClientSideKeypair keypair = this.keypairStore.getSnapshot().getKeypair(subscriptionId);
+        if(keypair == null) {
+            ResponseUtil.error(rc, 404, "Failed to find a keypair for subscription id: " + subscriptionId);
+            return;
+        }
+
+        JsonObject jo = toJsonWithPrivateKey(keypair);
+        rc.response()
+                .putHeader(HttpHeaders.CONTENT_TYPE, "application/json")
+                .end(jo.encode());
+    }
+
+    private JsonObject toJson(ClientSideKeypair keypair) {
+        JsonObject jo = new JsonObject();
+        jo.put("subscription_id", keypair.getSubscriptionId());
+        jo.put("public_key", keypair.encodePublicKeyToString());
+        jo.put("site_id", keypair.getSiteId());
+        jo.put("contact", keypair.getContact());
+        jo.put("created", keypair.getCreated().getEpochSecond());
+        jo.put("disabled", keypair.isDisabled());
+        return jo;
+    }
+
+    private JsonObject toJsonWithPrivateKey(ClientSideKeypair keypair) {
+        JsonObject jo = new JsonObject();
```

Comment:
can we call toJson() from here?

Reply:
Refactored to remove duplicated code but maintain consistent json orderering across endpoints.